de.variantsync.core/src/test/java/de/variantsync/core/ASTTest.java

```diff
@@ -0,0 +1,171 @@
+package de.variantsync.core;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import de.variantsync.core.ast.AST;
+import de.variantsync.core.ast.LineGrammar;
+
+/**
+ * Here you can find the unit tests for the AST data structure.
+ *
+ * @author eric
+ */
+public class ASTTest {
+
+	AST<LineGrammar, String> root;
+	AST<LineGrammar, String> mainDir;
+	AST<LineGrammar, String> testDir;
+	AST<LineGrammar, String> mainJava;
+	int lineIndex = 0; // only for toString testing
+	final int INITIAL_AST_SIZE = 11;
+	final int INITIAL_TOSTRING_ROWS = 10;
+
+	@Before
+	public void setup() {
+		root = new AST<>(LineGrammar.Directory, "src");
+		mainDir = new AST<>(LineGrammar.Directory, "main");
+		testDir = new AST<>(LineGrammar.Directory, "test");
+		mainJava = new AST<>(LineGrammar.TextFile, "Main.java");
+		final AST<LineGrammar, String> emptyJava = new AST<>(LineGrammar.TextFile, "Empty.java");
+		final AST<LineGrammar, String> emptyTestJava = new AST<>(LineGrammar.TextFile, "EmptyTest.java");
+		root.addChild(testDir);
+		testDir.addChild(emptyTestJava);
+		root.addChild(mainDir);
+		mainDir.addChild(mainJava);
+		mainDir.addChild(emptyJava);
+
+		mainJava.addChildren(
+				Arrays.asList(new AST<>(LineGrammar.Line, "public class Main {"), new AST<>(LineGrammar.Line, "    public static void main(String[] args)"),
+						new AST<>(LineGrammar.Line, "        System.out.println(\"Hello World\");"), new AST<>(LineGrammar.Line, "    }"),
+						new AST<>(LineGrammar.Line, "}")));
+	}
+
+	@Test
+	public void toStringOnInitialTest() {
+		final String[] lines = root.toString().split(" ");
+		assertEquals(INITIAL_TOSTRING_ROWS, lines.length);
+		// this loop skips the names of the variable and only checks their value
+		for (int i = 2; i < lines.length; i = i + 2) {
+			if (lines[i].charAt(lines[i].length() - 1) == ',') {
+				lines[i] = lines[i].substring(0, lines[i].length() - 1);
+			}
+			switch (i) {
+			case 2:
+				assertEquals(root.getId().getMostSignificantBits(), Long.parseLong(lines[i]));
+				break;
+			case 4:
+				assertEquals(root.getType().toString(), lines[i]);
+				break;
+			case 6:
+				assertEquals(root.getValue().toString(), lines[i]);
+				break;
+			case 8:
+				assertEquals(root.getSubtrees().size(), Integer.parseInt(lines[i]));
+				break;
+			default:
+				throw new IllegalArgumentException(String.format("toStringOnInitialTest has incorrect lines size: %d", i));
+			}
+		}
+	}
+
+	@Test
+	public void addOnInitialTest() {
```

Comment:
Isn't this rather testing if the addChild method respects the isValidChild method of the grammar correctly? Could you adapt the name to reflect that?